libraries/AP_Winch/AP_Winch_Servo.cpp

```diff
@@ -0,0 +1,92 @@
+#include <AP_Winch/AP_Winch_Servo.h>
+
+extern const AP_HAL::HAL& hal;
+
+void AP_Winch_Servo::init(const AP_WheelEncoder* wheel_encoder)
+{
+    _wheel_encoder = wheel_encoder;
+
+    // set servo output range
+    SRV_Channels::set_angle(SRV_Channel::k_winch,  1000);
+}
+
+void AP_Winch_Servo::update()
+{
+    // return immediately if no servo is assigned to control the winch
+    if (!SRV_Channels::function_assigned(SRV_Channel::k_winch)) {
+        return;
+    }
+
+    // return immediately if no wheel encoder
+    if (_wheel_encoder == nullptr) {
+        return;
+    }
+
+    // if not doing any control output trim value
+    if (config.state == AP_Winch::STATE_RELAXED) {
+        SRV_Channels::set_output_limit(SRV_Channel::k_winch, SRV_Channel::SRV_CHANNEL_LIMIT_TRIM);
+        return;
+    }
+
+    // calculate dt since last iteration
+    uint32_t now = AP_HAL::millis();
+    float dt = (now - last_update_ms) / 1000.0f;
+    if (dt > 1.0f) {
+        dt = 0.0f;
+    }
+    last_update_ms = now;
+
+    // calculate latest rate
+    float distance = _wheel_encoder->get_distance(0);
+    float rate = 0.0f;
+    if (is_positive(dt)) {
+        rate = (distance - _last_distance) / dt;
+    }
+    _last_distance = distance;
+
+    // update distance from wheel encoder
+    config.length_curr = distance;
+
+    // if doing position control, calculate position error to desired rate
+    float rate_desired = 0.0f;
+    if (config.state == AP_Winch::STATE_POSITION) {
+        float position_error = config.length_desired - config.length_curr;
+        rate_desired = constrain_float(position_error * config.pos_p, -config.rate_desired, config.rate_desired);
+    }
+
+    // if doing rate control, set desired rate
+    if (config.state == AP_Winch::STATE_RATE) {
+        rate_desired = config.rate_desired;
+    }
+
+    // calculate rate error and pass to pid controller
+    float rate_error = rate_desired - rate;
+    config.rate_pid.set_input_filter_all(rate_error);
+
+    // get p
+    float p = config.rate_pid.get_p();
+
+    // get i unless winch hit limit on last iteration
+    float i = config.rate_pid.get_integrator();
+    if (((is_negative(rate_error) && !limit_low) || (is_positive(rate_error) && !limit_high))) {
```

Comment:
I think I'm missing something here, but if on a loop through (with lets say a large integrated value, but a fairly low I grain, we could hit the output limit on one round because the integrator has built up high, so we output a max value value. On the next loop through the limit_high will be set, so we will instead put out the I gain. This is potentially small enough to drop us below the maximum output so we will only output a less then max value this update and unset the flag. The next update however we will go back to maxing it out.

Basically I think I'm seeing a possible control induced pulsing of the winch when operating near limits. I could easily be missing something about how the underlying controller is actually behaving though.

Reply:
In case it's not clear the "I gain" (i.e. "config.rate_pid.get_integrator()") will be the integrator value from the previous iteration in which the output hit the limit.  So the final output would only drop if the actual error has dropped.

Hopefully the integrator output won't be changing that quickly so hopefully the result of this is that the integrator stops building up at moments when the output is hitting the limits.

We do something very similar in Copter's attitude control.  I've tried to explain this on page 6 of this power point: https://www.dropbox.com/s/wleq7imwtgo468d/Randy_Elab_Offsite_Aug2017.pptx?dl=0

Reply:
Ah, okay I thought "I gain" was the actual tuning parameter of the I term, not the previous I value :)